Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp

```diff
@@ -0,0 +1,148 @@
+
+/// Algorithm's version for identification. @see Algorithm::version
+int CreateMonteCarloWorkspace::version() const { return 1; }
+
+/// Algorithm's category for identification. @see Algorithm::category
+const std::string CreateMonteCarloWorkspace::category() const { return "Simulation"; }
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+const std::string CreateMonteCarloWorkspace::summary() const {
+  return "Creates a randomly simulated workspace by sampling from the probability distribution of input data.";
+}
+
+//----------------------------------------------------------------------------------------------
+/** Initialize the algorithm's properties.
+ */
+void CreateMonteCarloWorkspace::init() {
+  auto mustBePositive = std::make_shared<Mantid::Kernel::BoundedValidator<int>>();
+  mustBePositive->setLower(0);
+
+  declareProperty(std::make_unique<WorkspaceProperty<MatrixWorkspace>>("InputWorkspace", "", Direction::Input),
+                  "Input Workspace containing data to be fitted");
+  declareProperty("Seed", 32, mustBePositive,
+                  "Integer that initializes a random-number generator, good for reproducibility");
+  declareProperty(std::make_unique<WorkspaceProperty<MatrixWorkspace>>("OutputWorkspace", "", Direction::Output),
+                  "Name of output workspace.");
+}
+
+//----------------------------------------------------------------------------------------------
+
+Mantid::HistogramData::HistogramY CreateMonteCarloWorkspace::fillHistogramWithRandomData(const std::vector<double> &cdf,
+                                                                                         int numIterations,
+                                                                                         std::mt19937 &gen) {
+  Mantid::HistogramData::HistogramY outputY(cdf.size(), 0.0);
+  std::uniform_real_distribution<> dis(0.0, 1.0);
+  for (int i = 0; i < numIterations; ++i) {
+    double randomNum = dis(gen);
+    // Find the bin corresponding to the random number in the CDF
+    auto it = std::lower_bound(cdf.begin(), cdf.end(), randomNum);
+    size_t index = std::distance(cdf.begin(), it);
+
+    // Ensure the index is within bounds
+    if (index < outputY.size()) {
+      outputY[index] += 1.0;
+    }
+  }
+  return outputY;
+}
+
+std::vector<double> CreateMonteCarloWorkspace::computeNormalizedCDF(const Mantid::HistogramData::HistogramY &yData) {
+  std::vector<double> cdf(yData.size());
+  std::partial_sum(yData.begin(), yData.end(), cdf.begin());
+  double total_counts = cdf.back();
+  // Normalize the CDF
+  std::transform(cdf.begin(), cdf.end(), cdf.begin(), [total_counts](double val) { return val / total_counts; });
+  return cdf;
+}
+
+int CreateMonteCarloWorkspace::computeNumberOfIterations(const Mantid::HistogramData::HistogramY &yData) {
+  double total_counts = std::accumulate(yData.begin(), yData.end(), 0.0);
+  return static_cast<int>(std::round(total_counts));
```

Comment:
What if this rounded to 0? I suppose you'd get 0s in the data, this might be confusing? Perhaps at least throw a warning?
Would it make more sense to add the number of MC events as an input parameter? Say for example you want to see how counting times/stats affect optimiser performance, I think to do this you would need to scale the input workspace which seems a  bit clunky.

Reply:
I think having the parameter to specify number of MC events is good, and the default could be the integral of the input.

Reply:
I commited the suggestion above, please have a look and let me know. The scaling function is not perfect, if the number of MC events is too little, the new workspace appears to be a flat line when overplotted onto the original workspace. I am open to any suggestions on ways to improve it.